gitlab_runner_config.py

```diff
@@ -62,111 +190,46 @@ def which(cmd):
-            }
-        )
-
-        request = Request(url, data=data.encode(), method="POST")
-        response = urllib.request.urlopen(request)
-        if response.getcode() == 201:
-            return json.load(response)
-        else:
-            print("Registration for {runner_type} failed".format(runner_type))
-            sys.exit(1)
-    except HTTPError as e:
-        print(
-            "Error registering runner {runner} with tags {tags}: {reason}".format(
-                runner=runner_type, tags=",".join(tags), reason=e.reason
-            )
-        )
-        sys.exit(1)
-
-
-def delete_runner(base_url, runner_token):
-    """Delete an existing runner"""
-
-    try:
-        url = urljoin(base_url, "runners")
-        data = urlencode(
-            {
-                "token": runner_token,
-            }
-        )
-
-        request = Request(url, data=data.encode(), method="DELETE")
-        response = urllib.request.urlopen(request)
-        if response.getcode() == 204:
-            return True
-        else:
-            print("Deleting runner with id failed")
-            sys.exit(1)
-    except HTTPError as e:
-        print("Error deleting runner: {reason}".format(reason=e.reason))
-        sys.exit(1)
-
+def read_runner(templates, path, clients):
+    """Reads a runner file and generates runner configurations"""
+    data_file = os.path.join(templates, path)
+    runners = {}
+    with open(data_file, "r") as fh:
+        runner_config = json.load(fh)
+        for runner_type, data in runner_config.items():
+            data = data or {}
+            name = data.get("name", "")
+            if name in runners:
+                print(f"Duplicate runner found for {name}")
+                sys.exit(1)
+            url = data.get("url", "")
+            # TODO: tokens in the runner configs for multiple gitlab urls, separate?
+            admin_token = data.get("admin_token", "")
+            access_token = data.get("access_token", "")
+            # removing admin/access tokens for future writing out to config.toml
+            del data["admin_token"]
+            del data["access_token"]
+
+            if url not in clients:
+                clients[url] = gitlab_client(url, admin_token, access_token)
+            client = clients[url]
+
+            runner_type = data.get("executor", "")
+            gitlab_tags = client.list_gitlab_tags()
+            config = data
+            if runner_type not in gitlab_tags:
+                # no config template tags in common with Gitlab, register runners
+                # for all the tags pulled from the template.
+                config = client.register_runner(
+                    runner_type,
+                    generate_tags(runner_type=runner_type),
+                )
+            else:
+                token = data.get("token", "")
+                # TODO: changed no longer needed?
+                config, _ = client.update_runner_token(token, runner_type)
```

Comment:
I think your TODO is right. The token should come back in the request to list runners with this host's tags so it can just get pulled from `data` as you're doing.